Sofa/Component/Constraint/Lagrangian/Model/src/sofa/component/constraint/lagrangian/model/FixedLagrangianConstraint.h

```diff
@@ -0,0 +1,96 @@
+#pragma once
+#include <sofa/component/constraint/lagrangian/model/config.h>
+
+#include <sofa/core/behavior/Constraint.h>
+#include <sofa/core/behavior/ConstraintResolution.h>
+#include <sofa/core/behavior/MechanicalState.h>
+#include <sofa/core/behavior/OdeSolver.h>
+#include <sofa/core/topology/TopologySubsetIndices.h>
+
+namespace sofa::component::constraint::lagrangian::model
+{
+
+template< class DataTypes >
+class FixedLagrangianConstraint : public core::behavior::Constraint<DataTypes>
+{
+public:
+    SOFA_CLASS(SOFA_TEMPLATE(FixedLagrangianConstraint,DataTypes), SOFA_TEMPLATE(core::behavior::Constraint,DataTypes));
+
+    typedef typename DataTypes::VecCoord VecCoord;
+    typedef typename DataTypes::VecDeriv VecDeriv;
+    typedef typename DataTypes::Coord Coord;
+    typedef typename DataTypes::Deriv Deriv;
+    typedef typename DataTypes::MatrixDeriv MatrixDeriv;
+    typedef typename Coord::value_type Real;
+    typedef typename core::behavior::MechanicalState<DataTypes> MechanicalState;
+    typedef typename core::behavior::Constraint<DataTypes> Inherit;
+
+    typedef typename DataTypes::MatrixDeriv::RowIterator MatrixDerivRowIterator;
+    typedef core::objectmodel::Data<VecCoord>		DataVecCoord;
+    typedef core::objectmodel::Data<VecDeriv>		DataVecDeriv;
+    typedef core::objectmodel::Data<MatrixDeriv>    DataMatrixDeriv;
+
+    typedef sofa::core::topology::TopologySubsetIndices SetIndex;
+
+
+protected:
+
+    std::vector<unsigned int> m_cid;
+    std::vector<Deriv> m_prevForces;
+
+    SetIndex d_indices;
+    Data<bool> d_fixAll;
+
+
+
+    FixedLagrangianConstraint(MechanicalState* object = nullptr);
+    virtual ~FixedLagrangianConstraint() {}
+
+
+    virtual type::vector<std::string> getConstraintIdentifiers() override final
+    {
+        type::vector<std::string> ids = getFixedIdentifiers();
+        ids.push_back("Bilateral");
+        return ids;
+    }
+
+    virtual type::vector<std::string> getFixedIdentifiers(){ return {}; }
+
+
+
+public:
+    void init() override;
+    void buildConstraintMatrix(const core::ConstraintParams* cParams, DataMatrixDeriv &c_d, unsigned int &cIndex, const DataVecCoord &x) override;
+    void getConstraintViolation(const core::ConstraintParams* cParams, linearalgebra::BaseVector *resV, const DataVecCoord &x, const DataVecDeriv &v) override;
+    void getConstraintResolution(const core::ConstraintParams *, std::vector<core::behavior::ConstraintResolution*>& resTab, unsigned int& offset) override;
+
+private:
+    void doBuildConstraintLine(MatrixDeriv &c, unsigned int &cIndex, unsigned int lineNumber);
+    void doGetSingleConstraintViolation(linearalgebra::BaseVector *resV, const DataVecCoord * freePose, const DataVecCoord * restPose,unsigned int lineNumber);
+    void doGetSingleConstraintResolution(std::vector<core::behavior::ConstraintResolution*>& resTab, unsigned int& offset, unsigned int lineNumber);
+
+};
+
+
```

Comment:
The `extern template` thing is missing

Reply:
Yeap, I actually didn't know why it was needed. Is it required for the cuda implementation or such extension ?

Reply:
it's a trick to reduce compilation time in case your class is used elsewhere. With the `extern template` keywords, the compiler says "I have generated, or I will, the class from the listed template. Trust me, you don't have to generate it again"

Reply:
Ah yes, I see ! Thank's